### PR TITLE
[MM-29296] model/user: add lock to source

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary

This PR adds a lock to passwordRandomSource to be able to use safely across multiple goroutines. 

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-29296

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: model/user.go [hunk 1]
```diff
@@ -884,24 +885,39 @@ func UsersWithGroupsAndCountFromJson(data io.Reader) *UsersWithGroupsAndCount {
...
 
-var passwordRandomSource = rand.NewSource(time.Now().Unix())
+type lockedSource struct {
+	mu  sync.Mutex
+	src rand.Source
+}
+
+func (r *lockedSource) Intn(n int) int {
+	r.mu.Lock()
+	rn := rand.New(r.src)
```
Comment (streamer45) [inline model/user.go:895]: I would expect to create the generator (`rand.*Rand`) only once and use it as needed.
Here instead we create a new one every time we need to generate a random integer. Is there a reason for that?

Comment (isacikgoz) [reply model/user.go:895]: No specific reason, it was used like this way so, I adopted that approach. It can be removed.

Comment (streamer45) [reply model/user.go:895]: Before the generator was created only once on top of `GeneratePassword` and then used.
Here we are essentially creating a brand new one for each `Intn` call.

Comment (isacikgoz) [reply model/user.go:895]: Yeah 🤦

File: model/user.go [hunk 2]
```diff
@@ -884,24 +885,38 @@ func UsersWithGroupsAndCountFromJson(data io.Reader) *UsersWithGroupsAndCount {
...
+func (r *lockedRand) Intn(n int) int {
+	r.mu.Lock()
+	m := r.rn.Intn(n)
+	r.mu.Unlock()
+	return m
+}
+
+var passwordRandom = lockedRand{
+	rn: rand.New(rand.NewSource(time.Now().Unix())),
+	mu: sync.Mutex{},
```
Comment (streamer45) [inline model/user.go:902]: I don't think this is necessary since it's a value. Same for the one in the test case below.

**Review by streamer45 — APPROVED** [review]
Looks good :+1:

**Comment by isacikgoz** [comment]
@lindalumitchell this is a fix for a data race in the code, a smoke test with bulk import would be fine.

**Comment by agnivade** [comment]
@lindalumitchell - gentle ping on this. :)

**Review by lindalumitchell — APPROVED** [review]
Goodness, sorry @agnivade and thank you for the ping! The old-style test server doesn't seem to be building; I'm okay with getting this in and testing it after merge. Sorry again for the delay!

**Pull request merged**